### PR TITLE
fix: Fix the recommended application display exception

### PR DESCRIPTION
**Pull request by ssongliu**

### CODE REVIEW

File: core/app/service/app_launcher.go [hunk 1]
```diff
@@ -54,7 +54,19 @@ func (u *LauncherService) ChangeShow(req dto.SettingUpdate) error {
...
+		launcherMap[item.Key] = struct{}{}
+		list = append(list, item.Key)
+	}
+	launcherData := struct {
+		Keys []string
+	}{Keys: list}
+	itemData, _ := json.Marshal(launcherData)
 	_, _ = proxy_local.NewLocalClient("/api/v2/dashboard/app/launcher/sync", http.MethodPost, bytes.NewReader((itemData)))
 	_ = xpack.RequestToAllAgent("/api/v2/dashboard/app/launcher/sync", http.MethodPost, bytes.NewReader((itemData)))
 }
```
Comment (wanghe-fit2cloud) [inline core/app/service/app_launcher.go:72]: The code snippet you provided is designed to sync launcher information with agents. It uses JSON marshalling to serialize launcher data for sending over an HTTP POST request to local client and all agents. Here are some points to consider:

1. **Simplified Json Marshaling**:
   - The current implementation serializes the entire `launchers` slice directly into JSON without considering specific keys for deduplication.
   - Simplifying the logic will improve efficiency, especially if there are a large number of entries.

2. **Deduplication Logic**:
   - Adding deduplication by checking against a map (`launcherMap`) significantly improves performance, especially when dealing with many items (e.g., thousands).
   - Ensure that this logic works as intended, especially if launchers can have duplicate keys based on certain criteria.

3. **Error Handling**:
   - There are no explicit error checks or handling within the loop, which could lead to silent failures if there are issues during iteration or mapping conversion.
   - Consider adding more comprehensive error handling, such as logging errors or retrying failed requests.

4. **Security Considerations**:
   - While not explicitly mentioned here, ensure that sensitive information is handled securely, especially around serialization and network communication.

5. **Testing**:
   - Before deploying changes, create comprehensive unit tests to cover edge cases, including duplicate items, empty slices, etc.

Overall, the use of the map for deduplication simplifies the process but requires careful validation.

File: agent/app/service/dashboard.go [hunk 2]
```diff
@@ -47,10 +47,10 @@ func NewIDashboardService() IDashboardService {
...
-func (u *DashboardService) Sync(req []dto.AppLauncherSync) error {
+func (u *DashboardService) Sync(req dto.AppLauncherSync) error {
 	var launchers []model.AppLauncher
-	for _, item := range req {
-		launchers = append(launchers, model.AppLauncher{Key: item.Key})
+	for _, item := range req.Keys {
+		launchers = append(launchers, model.AppLauncher{Key: item})
 	}
 	return launcherRepo.SyncAll(launchers)
 }
```
Comment (wanghe-fit2cloud) [inline agent/app/service/dashboard.go:56]: The provided code looks mostly correct, but there are a few minor improvements and optimizations:

1. **Use `req.Keys` instead of `item.Key`**: Ensure that you're iterating over the slice correctly.

2. **Error Handling**: The current implementation doesn't return an error if appending to `launchers` fails. Consider adding proper handling for errors during this operation.

3. **Performance Optimization**: If inserting thousands of items into the database is a common operation, consider using batch inserts to improve performance.

Here's a revised version with these considerations in mind:

```go
import (
	"errors"

	type DashboardService struct{}

	type IDashboardService interface {
		Sync(req dto.AppLauncherSync) error
	}

	type AppLauncher struct {
		Key string
	}
)

type Dashboardservice struct{
	appLauncherRepo model.AppLauncherRepository
}

func NewIDashboardService(app LauncherReposiroty ) *Dashboardservice {

	return &Dashboardservice{appLauncherRepo: app}
}

func (u *Dashboadservice) Sync(req dto.AppLauncherSync) error {
    var launchers []AppLauncher

	for _, item := range req.Keys {
        l := AppLauncher{Key: item}
        launchers = append(launchers, l)
	}

	err := u.appLauncherRepo.SyncAll(launchers)
	if err != nil {
		err = fmt.Errorf("error syncing launchers: %v", err)
	}
	
	return err
}
```

### Key Improvements:
- **Consistent Use of Keys**: Ensured all keys from `dto.AppLauncherSync.Keys` are processed properly.
- **Error Handling**: Added a specific error message when encountering issues with `launcherRepo.SyncAll`.
- **Type Clarity**: Simplified internal structure by introducing a separate `AppLauncher` struct to hold key fields directly.

File: agent/app/api/v2/dashboard.go [hunk 3]
```diff
@@ -62,7 +62,7 @@ func (b *BaseApi) LoadAppLauncherOption(c *gin.Context) {
 }
 
 func (b *BaseApi) SyncAppLauncher(c *gin.Context) {
-	var req []dto.AppLauncherSync
+	var req dto.AppLauncherSync
 	if err := helper.CheckBindAndValidate(&req, c); err != nil {
 		return
 	}
```
Comment (wanghe-fit2cloud) [inline agent/app/api/v2/dashboard.go:68]: The provided code snippet has no obvious irregularities or major issues at this time. However, there is one suggestion for improvement:

```go
var req dto.AppLauncherSync
```

You can make it explicit that `req` should be a pointer type to avoid unnecessary copying of data in the function and allow for more efficient use within other parts of the application if needed.

Here's the revised line:
```go
var req *dto.AppLauncherSync
```

This change will help reduce memory usage and maintainability of the code while allowing for safer manipulation of the struct fields internally if needed.
